oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/async/ElasticResultRowAsyncIterator.java

```diff
@@ -0,0 +1,332 @@
+            listeners.add(elasticFacetProvider);
+        }
+
+        return new ElasticQueryScanner(elasticRequestHandler, listeners);
+    }
+
+    /**
+     * Scans Elastic results asynchronously and notify listeners.
+     */
+    class ElasticQueryScanner implements ActionListener<SearchResponse> {
+
+        private static final int SMALL_RESULT_SET_SIZE = 10;
+        private static final int MEDIUM_RESULT_SET_SIZE = 100;
+        private static final int LARGE_RESULT_SET_SIZE = 1000;
+
+        private final Set<ElasticResponseListener> allListeners = new HashSet<>();
+        private final List<SearchHitListener> searchHitListeners = new ArrayList<>();
+        private final List<AggregationListener> aggregationListeners = new ArrayList<>();
+
+        private final QueryBuilder query;
+        private final String[] sourceFields;
+
+        // concurrent data structures to coordinate chunks loading
+        private final AtomicBoolean anyDataLeft = new AtomicBoolean(false);
+
+        private int scannedRows = 0;
+        private boolean firstRequest = true;
+        private boolean fullScan = false;
+
+        // reference to the last document sort values for search_after queries
+        private Object[] lastHitSortValues;
+
+        // Semaphore to guarantee only one in-flight request to Elastic
+        private final Semaphore semaphore = new Semaphore(1);
+
+        ElasticQueryScanner(ElasticRequestHandler requestHandler,
+                            List<ElasticResponseListener> listeners) {
+            this.query = requestHandler.build();
+
+            final Set<String> sourceFieldsSet = new HashSet<>();
+            final AtomicBoolean needsAggregations = new AtomicBoolean(false);
+            final Consumer<ElasticResponseListener> register = (listener) -> {
+                allListeners.add(listener);
+                sourceFieldsSet.addAll(listener.sourceFields());
+                if (listener instanceof SearchHitListener) {
+                    searchHitListeners.add((SearchHitListener) listener);
+                    if (((SearchHitListener) listener).isFullScan()) {
+                        fullScan = true;
+                    }
+                }
+                if (listener instanceof AggregationListener) {
+                    aggregationListeners.add((AggregationListener) listener);
+                    needsAggregations.set(true);
+                }
+            };
+
+            listeners.forEach(register);
+            this.sourceFields = sourceFieldsSet.toArray(new String[0]);
+
+            final SearchSourceBuilder searchSourceBuilder = SearchSourceBuilder.searchSource()
+                    .query(query)
+                    // use a smaller size when the client asks for facets. This improves performance
+                    // when the client is only interested in insecure facets
+                    .size(needsAggregations.get() ? SMALL_RESULT_SET_SIZE : MEDIUM_RESULT_SET_SIZE)
```

Comment:
Are we using small dataset in case of aggregation even in case when user is asking for other fields also?

Reply:
Yes, this happens only for `insecure` facets. It could require one more call to Elastic in case the user is asking for other fields and the result set contains between 11 and 1100 rows. Regardless of the result set size, it would return facets fast. We could make it more sophisticated and ask for the `SMALL_RESULT_SET` when just insecure facets are requested. I am not sure how this scenario is frequent and if it worth adding more complexity.